### ISSUE TITLE
rapid broker RSS growth and slow job-manager response on large system

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#6114]
We caught another instance of rapid RSS growth (final RSS was >210GB) of the broker on elcap.

During this time, the flux-broker was using 400% CPU and I was able to capture a `perf` trace. Turns out ~23% of the broker's time was in the job-manager handling annotations. Another 20% was in the Fluxion qmanager sending annotations (we should open a corresponding issue in flux-sched)

```
   - ev_run                                                                    ▒
      - 22.69% ev_invoke_pending                                               ▒
         - 21.61% handle_cb                                                    ▒
            - 20.48% dispatch_message (inlined)                                ▒
               - 20.07% call_handler                                           ▒
                  - 19.91% alloc_response_cb                                   ▒
                     - 15.70% annotations_update_and_publish                   ▒
                        - 14.73% event_job_post_pack                           ▒
                           - 14.72% event_job_post_vpack                       ▒
                              - 13.10% event_job_post_deferred                 ▒
                                 - 11.74% event_job_process_entry              ▒
                                    - 10.51% journal_process_event             ▒
                                       - 9.77% flux_respond_pack               ▒
                                          + 9.76% flux_respond_vpack (inlined) ▒
```

and 

```
   - ev_run                                                                    ▒
      - 18.03% ev_invoke_pending                                               ▒
         - 14.91% qmanager_safe_cb_t::check_watcher_cb                         ▒
            - 14.71% qmanager_safe_cb_t::check_watcher_cb                      ▒
               - qmanager_cb_t::check_watcher_cb                               ▒
                  - 13.89% qmanager_cb_t::post_sched_loop                      ▒
                     + 12.24% schedutil_alloc_respond_annotate_pack            ▒
                     + 1.34% ?? (inlined)                                      ▒
```

It may be (partially verified by @trws and @milroy) that a bug in the qmanager, plus the new asynchronous code is resulting in many more annotation updates for jobs than before.

This doesn't fully explain the massive RSS growth though. However, it does give us a hint where to look. Even if the qmanager "bug" is fixed, we should still figure out why annotation updates are (possibly) chewing up massive amounts of RAM, perhaps we can do better here.

One experiment would be to just add a bit of code to sched-simple that sends annotation updates for all jobs in a check watcher (or even in a loop)

**Comment by grondo** [comment]
Unsurprisingly job-list is also busy:
```
-   12.98%     0.00%  job-list         libc-2.28.so                [.] __GI___c▒
     __GI___clone (inlined)                                                    ▒
     start_thread                                                              ▒
     module_thread                                                             ▒
     mod_main                                                                  ▒
     flux_reactor_run                                                          ▒
     ev_run                                                                    ▒
   - ev_run                                                                    ▒
      - 11.29% ev_invoke_pending                                               ▒
         - 7.94% job_events_journal_continuation                               ▒
            - 5.12% flux_rpc_get_unpack                                        ▒
               - flux_rpc_get_vunpack (inlined)                                ▒
                  - 5.10% flux_msg_vunpack                                     ▒
                     - 4.89% json_loads                                        ▒
```

**Comment by grondo** [comment]
I was able to reproduce a large and fast broker RSS growth by instrumenting sched-simple to send 10K annotation updates per job, simulating what we think is happening with Fluxion qmanger.

With this, I was able to run the reproducer under a flux built with `--with-tcmalloc` and using heaptrace. Here's the result between the first heap dump and awhile later

[heap.pdf](https://github.com/user-attachments/files/16277112/heap.pdf)

It appears that all the memory is going into the job eventlogs. One guess is that even though annotation updates are JOURNAL_ONLY, they are still kept indefinitely in the in-memory copy of the eventlog in the job manager?

**Comment by garlick** [comment]
We don't keep the journal but we do keep the individual job eventlogs in job->eventlog.

**Comment by garlick** [comment]
This seems to help
```diff
diff --git a/src/modules/job-manager/event.c b/src/modules/job-manager/event.c
index dbdda8186..6985d5c65 100644
--- a/src/modules/job-manager/event.c
+++ b/src/modules/job-manager/event.c
@@ -814,9 +814,11 @@ int event_job_process_entry (struct event *event,
         }
     }
 
-    if (json_array_append (job->eventlog, entry) < 0) {
-        errno = ENOMEM;
-        return -1;
+    if (!(flags & EVENT_NO_COMMIT)) {
+        if (json_array_append (job->eventlog, entry) < 0) {
+            errno = ENOMEM;
+            return -1;
+        }
     }
 
     if (journal_process_event (event->ctx->journal,
```

although I still see (slower) RSS growth in my testing, probably due to the message backlog created for the job manager.  It doesn't shrink if I stop the scheduler and let it catch up though which is a little concerning.  OTOH if I start it up again there is a delay before it starts edging up which makes me think malloc just hasn't let go of that memory and it is getting reused.

**Comment by grondo** [comment]
Ok, I just tried the same patch:

```diff
diff --git a/src/modules/job-manager/event.c b/src/modules/job-manager/event.c
index dbdda8186..eef5280f9 100644
--- a/src/modules/job-manager/event.c
+++ b/src/modules/job-manager/event.c
@@ -814,7 +814,8 @@ int event_job_process_entry (struct event *event,
         }
     }
 
-    if (json_array_append (job->eventlog, entry) < 0) {
+    if (!(flags & EVENT_NO_COMMIT)
+        && json_array_append (job->eventlog, entry) < 0) {
         errno = ENOMEM;
         return -1;
     }
```

I see RSS growing to a point in my test (808M) but it stops growing after that (in my previous run the RSS reached several GB very quickly).

I can capture another heap trace to track the memory. If malloc is just not freeing it to the OS then heaptrace will not see much difference in the used memory between two dumps I think.

**Comment by garlick** [comment]
Fixed by #6115 - reopen if there is more to do here that I missed.
